core/src/main/java/com/linecorp/armeria/internal/common/stream/DefaultHttpDeframer.java

```diff
@@ -0,0 +1,301 @@
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.stream.AbortedStreamException;
+import com.linecorp.armeria.common.stream.CancelledSubscriptionException;
+import com.linecorp.armeria.common.stream.DefaultStreamMessage;
+import com.linecorp.armeria.common.stream.HttpDecoder;
+import com.linecorp.armeria.common.stream.HttpDecoderOutput;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.stream.SubscriptionOption;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * The default HTTP deframer implementation.
+ */
+@UnstableApi
+public final class DefaultHttpDeframer<T> extends DefaultStreamMessage<T> implements HttpDecoderOutput<T> {
+
+    private static final SubscriptionOption[] EMPTY_OPTIONS = {};
+    private static final SubscriptionOption[] POOLED_OBJECTS_OPTIONS = { WITH_POOLED_OBJECTS };
+    private static final SubscriptionOption[] NOTIFY_CANCELLATION_OPTIONS = { NOTIFY_CANCELLATION };
+
+    private final HttpMessageSubscriber subscriber = new HttpMessageSubscriber();
+
+    private final HttpDecoder<T> handler;
+    private final ByteBufDecoderInput input;
+    private final Function<? super HttpData, ? extends ByteBuf> byteBufConverter;
+    private final StreamMessage<? extends HttpObject> publisher;
+
+    @Nullable
+    private RequestHeaders requestHeaders;
+    @Nullable
+    private Subscription upstream;
+
+    private boolean handlerProduced;
+    private boolean sawLeadingHeaders;
+    private boolean initialized;
+    private boolean askedUpstreamForElement;
+    private boolean cancelled;
+
+    /**
+     * Returns a new {@link DefaultHttpDeframer} with the specified {@link HttpDecoder} and
+     * {@link ByteBufAllocator}.
+     */
+    public DefaultHttpDeframer(StreamMessage<? extends HttpObject> streamMessage,
+                               HttpDecoder<T> handler, ByteBufAllocator alloc) {
+        this(streamMessage, handler, alloc, HttpData::byteBuf);
+    }
+
+    /**
+     * Returns a new {@link DefaultHttpDeframer} with the specified {@link HttpDecoder},
+     * {@link ByteBufAllocator} and {@code byteBufConverter}.
+     */
+    public DefaultHttpDeframer(StreamMessage<? extends HttpObject> streamMessage,
+                               HttpDecoder<T> handler, ByteBufAllocator alloc,
+                               Function<? super HttpData, ? extends ByteBuf> byteBufConverter) {
+        publisher = requireNonNull(streamMessage, "streamMessage");
+        this.handler = requireNonNull(handler, "handler");
+        input = new ByteBufDecoderInput(requireNonNull(alloc, "alloc"));
+        this.byteBufConverter = requireNonNull(byteBufConverter, "byteBufConverter");
+        if (publisher instanceof HttpRequest) {
+            requestHeaders = ((HttpRequest) publisher).headers();
```

Comment:
Shouldn't we add `sawLeadingHeaders = true`?

Reply:
The `requestHeaders` is delivered via `HttpMessageSubscriber.onNext()`. 
`sawLeadingHeaders` will be set to `true` if the `requestHeaders` is actually received.

Reply:
Ah I missed that. 😅